[AI Gallery] Databricks DBRX LLM (#3379)
skypilot-org/skypilot: llm/dbrx/README.md

# Databricks DBRX: A State-of-the-Art Open LLM

<p align="center">
  <img src="https://www.databricks.com/en-blog-assets/static/2fe1a0af1ee0f6605024a810b604079c/dbrx-blog-header-optimized.png" alt="DBRX Blog Header" height="200">
</p>

[DBRX](https://www.databricks.com/blog/introducing-dbrx-new-state-art-open-llm) is an open, general-purpose LLM created by Databricks. It uses a mixture-of-experts (MoE) architecture with 132B total parameters of which 36B parameters are active on any input.

In this recipe, you will serve `databricks/dbrx-instruct` on your own infra  -- existing Kubernetes cluster or cloud VMs -- with one command.

## Prerequisites

- Go to the [HuggingFace model page](https://huggingface.co/databricks/dbrx-instruct) and request access to the model `databricks/dbrx-instruct`.
- Check that you have installed SkyPilot ([docs](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)).
- Check that `sky check` shows clouds or Kubernetes are enabled.

## SkyPilot YAML

See the full YAML file [here](https://github.com/skypilot-org/skypilot/tree/master/llm/dbrx/dbrx.yaml).

> 

Reviewer: Could we add the YAML inline and collapsed here?

Author: Yep, done.

## Serving DBRX: single instance

> Launch a single spot instance ($4.1/hour up) to serve DBRX on your infra:

Reviewer: ```suggestion
Launch a single spot instance (starts from $4.1/hour) to serve DBRX on your infra:
```

> Launch a single spot instance ($4.1/hour up) to serve DBRX on your infra:

Reviewer 2: mention acc type for the price?

```console

> HF_TOKEN=xxx sky launch dbrx.yaml -c dbrx --env HF_TOKEN

Reviewer 2: ```suggestion
sky launch dbrx.yaml -c dbrx --env HF_TOKEN=xxx
```

Author: The previous way is more secure.

Reviewer 2: SG

(task, pid=17433) INFO:     Application startup complete.
(task, pid=17433) INFO:     Uvicorn running on http://0.0.0.0:8081 (Press CTRL+C to quit)
...
(task, pid=17433) Running on local URL:  http://127.0.0.1:8811
(task, pid=17433) Running on public URL: https://xxxxxxxxxx.gradio.live
...
(task, pid=17433) INFO 03-28 04:32:50 metrics.py:218] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.0%, CPU KV cache usage: 0.0%
```
:tada: **Congratulations!** :tada: You have now launched the DBRX Instruct LLM on your infra.

You can play with the model via
- Standard OpenAPI-compatible endpoints (e.g., `/v1/chat/completions`)
- Gradio UI (automatically launched)

To curl `/v1/chat/completions`:
```console
IP=$(sky status --ip dbrx)

curl http://$IP:8081/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "databricks/dbrx-instruct",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "Who are you?"
      }
    ]
  }'
```

> To use the Gradio UI, open the URL shown in the logs:
> ```console
> (task, pid=17433) Running on local URL:  http://127.0.0.1:8811
> (task, pid=17433) Running on public URL: https://xxxxxxxxxx.gradio.live
> ```

Reviewer 2: ```suggestion
To use the Gradio UI, open the URL shown in the logs:
```console
(task, pid=17433) Running on public URL: https://xxxxxxxxxx.gradio.live
```
```
should we only include the gradio URL to reduce confusion

<img src="https://i.imgur.com/BZszerX.png" alt="Gradio UI serving DBRX" style="height: 600px;">

> 

Reviewer: Can we center this image?

Reviewer: not strongly, depends on time: can we make it a gif if possible to show the processing speed.

Reviewer: also, not strongly: if time permits, we may want to find a representative lines that can show why DBRX being "SOTA"

To stop the instance:
```console
sky stop dbrx
```

To shut down all resources:
```console
sky down dbrx
```

## Serving DBRX: scaling up with SkyServe

> After playing with the model, you can deploy the model with autoscaling and load-balancing using SkyServe.

Reviewer 2: mention now we could handle spot preemption and reduce downtime?

Reviewer 2: oh just saw it in latter position. not strongly, but feeling like we could mention it in 1 sentence to emphasize why users what to use skyserve

With no change to the YAML, launch a fully managed service on your infra:
```console
HF_TOKEN=xxx sky serve up dbrx.yaml -n dbrx --env HF_TOKEN
```

Wait until the service is ready:
```console
watch -n10 sky serve status dbrx
```

<details>
<summary>Example outputs:</summary>

```console
Services
NAME  VERSION  UPTIME  STATUS  REPLICAS  ENDPOINT
dbrx  1        35s     READY   2/2       xx.yy.zz.100:30001

Service Replicas
SERVICE_NAME  ID  VERSION  IP            LAUNCHED     RESOURCES                       STATUS  REGION
dbrx          1   1        xx.yy.zz.121  18 mins ago  1x GCP([Spot]{'A100-80GB': 4})  READY   us-east4
dbrx          2   1        xx.yy.zz.245  18 mins ago  1x GCP([Spot]{'A100-80GB': 4})  READY   us-east4
```
</details>

Get a single endpoint that load-balances across replicas:
```console
ENDPOINT=$(sky serve status --endpoint dbrx)
```

> **Tip:** SkyServe fully manages the lifecycle of your replicas. For example, if a spot replica is preempted, the controller will automatically replace it. This significantly reduces the operational burden while saving costs.

To curl the endpoint:
```console

> curl $ENDPOINT/v1/chat/completions \

Reviewer: ```suggestion
curl -L $ENDPOINT/v1/chat/completions \
```
